lib/tapioca/dsl/compilers/ffi.rb

```diff
@@ -0,0 +1,196 @@
+# frozen_string_literal: true
+
+begin
+  require "ffi"
+rescue LoadError
+  return
+end
+
+module Tapioca
+  module Dsl
+    module Compilers
+      # `Tapioca::Dsl::Compilers::FFI` generate types for functions and
+      # variables bound by Ruby-FFI for dynamically-linked native libraries.
+      # Ruby-FFI dynamically defines constants and methods at runtime. For
+      # example, given a module:
+      #
+      #   module MyModule
+      #     extend FFI::Library
+      #     ffi_lib 'mylib'
+      #
+      #     attach_function :func, [:int, :pointer], :void
+      #     attach_variable :var, :string
+      #   end
+      #
+      # This will result in the following methods being defined:
+      #
+      #   func, var, var=
+      #
+      class FFI < Compiler
+        def decorate
+          root.create_path(constant) do |mod|
+            constant.class_variables.each do |class_var|
+              candidate = constant.class_variable_get(class_var)
+              if candidate.is_a?(::FFI::Function)
+                method_name = class_var.to_s.gsub(/\A@*/, "")
```

Comment:
Accessing internal variables of FFI is risky. I think this is a good reason to add this compiler directly into ffi.

On the other hand I wonder, if we shouldn't make a public API to query this information and for other two cases below. Then I would prefer to store the compiler in another gem.

Reply:
It feels pretty hacky to me, I think new method(s) to query what is needed is much better.
For instance this logic is unlikely to work on the TruffleRuby backend of ffi.

Reply:
Thanks for your feedback both. All good points. I've pushed a change that exposes attached functions, getters and setters along with metadata in the following format (up for discussion):
- For getters and setters, a map from symbol (`:foo` or `:foo=`) to type class (e.g. `FFI::Type::STRING`)
- For functions, a map from symbol to an array of the shape `[param_types, result_type]` mimicking the arguments to `attach_function`, again using type classes.
- For variadic functions, the same with symbol `:varargs` tacked on the end of the `param_types`.

I'm a bit unhappy with `param_types` containing a mix of classes (`FFI::Type::String`) and the `:varargs` symbol. It makes me wonder if maybe the values should contain symbols instead (e.g. `:string`). However, we'd still have a mix when e.g. structure references are used so I'm not sure that would be an improvement. It would also force me to duplicate some of the type resolution code from FFI in the compiler.

Let me know what you think.

Reply:
PS I haven't tested the latest changes to the DSL compiler yet, I've added them only for illustration.

Reply:
It just occurred to me that since getters and setters _are_ functions, we might as well roll them into the `attached_functions` data, wdyt?